src/metrics/include/metrics/operation.hpp

```diff
@@ -0,0 +1,260 @@
+
+public:
+    using time_point = typename ClockSource::time_point;
+
+    explicit OperationContextT(v1::OperationImpl<ClockSource>& op)
+        : _op{std::addressof(op)}, _started{ClockSource::now()} {}
+
+    OperationContextT(OperationContextT<ClockSource>&& other) noexcept
+        : _op{std::move(other._op)},
+          _started{std::move(other._started)},
+          _event{std::move(other._event)},
+          _isClosed{std::exchange(other._isClosed, true)} {}
+
+    ~OperationContextT() {
+        if (!_isClosed) {
+            BOOST_LOG_TRIVIAL(warning)
+                << "Metrics not reported because operation '" << _op->getOpName()
+                << "' being run by actor '" << _op->getActorName()
+                << "' did not close with success() or failure().";
+        }
+    }
+
+    /**
+     * Increments the counter for the number of iterations. This method only needs to be called if
+     * an actor is periodically reporting its operations. By default an `iters = 1` value is
+     * automatically reported.
+     */
+    void addIterations(count_type iters) {
+        _event.iters += iters;
+    }
+
+    /**
+     * Increments the counter for the number of documents inserted, modified, deleted, etc.
+     */
+    void addDocuments(count_type ops) {
+        _event.ops += ops;
+    }
+
+    /**
+     * Increments the size in bytes of the number of documents inserted, etc.
+     */
+    void addBytes(count_type size) {
+        _event.size += size;
+    }
+
+    /**
+     * Increments the counter for the number of write errors, transient transaction errors, etc.
+     */
+    void addErrors(count_type errors) {
+        _event.errors += errors;
+    }
+
+    /**
+     * Report the operation as having succeeded.
+     *
+     * @warning After calling success() it is illegal to call any methods on this instance.
+     */
+    void success() {
+        reportOutcome(OutcomeType::kSuccess);
+    }
+
+    /**
+     * Report the operation as having failed. An `errors > 0` value won't be reported unless
+     * addErrors() has already been called.
+     *
+     * @warning After calling failure() it is illegal to call any methods on this instance.
+     */
+    void failure() {
+        reportOutcome(OutcomeType::kFailure);
+    }
+
+    /**
+     * Don't report the operation.
+     *
+     * @warning After calling discard() it is illegal to call any methods on this instance.
+     */
+    void discard() {
+        _isClosed = true;
+    }
+
+private:
+    void reportOutcome(OutcomeType outcome) {
+        auto finished = ClockSource::now();
+        _event.duration = finished - _started;
+        _event.outcome = outcome;
+
+        if (_event.iters == 0) {
+            // We default the event to represent a single iteration of a loop if addIterations() was
+            // never called.
+            _event.iters = 1;
+        }
+
+        _op->reportAt(finished, std::move(_event));
+        _isClosed = true;
```

Comment:
since we moved-from the event should we put ourselves into a moved-from state here as well? (basically `swap` this with a new instance)

if the user calls `.failure()` and `.success()` (or either more than once) what will the actual behavior be? (But otoh you've got good comments and `@warning`s on the various methods so not a huge deal if this turns out to be painful to change)

Reply:
> if the user calls `.failure()` and `.success()` (or either more than once) what will the actual behavior be?

I'm not sure `std::move(_event)` can realistically lead to anything other than the time series inserting a copy of `_event`. I suspect calling `success()` or `failure()` multiple times to report events as though multiple of them were started at the same time. How does Cedar handle effectively concurrent operations for the same (Actor, Operation, Thread) triplet?

> since we moved-from the event should we put ourselves into a moved-from state here as well? (basically `swap` this with a new instance)

What exactly are you trying to achieve? I'm curious if you're actually looking for `OperationContextT` to define a move assignment operator so the start time is able to be updated.

```cpp
OperationContextT& operator=(OperationContextT&& other) {
    _op = std::move(other._op);
    _started = std::move(other._started);
    _event = std::move(other._event);
    _isClosed = std::exchange(other._isClosed, true);
    return *this;
}
```

Reply:
> How does Cedar handle effectively concurrent operations for the same (Actor, Operation, Thread) triplet?

Cedar doesn't really care about thread numbers so it'd basically treat it the same as two threads reporting the same data at the same time.

> What exactly are you trying to achieve?

Designing types is hard especially given that we have an explicit move-from for only a single instance-var in a non-move operator/ctor. I get nervous whenever I see `std::move` on an instance-variable that doesn't result in the whole instance being in a moved-from state. I'm not sure if you need `operator=(&&)` as such (and it would force you to make some fields non-`const`).

I'm not sure what to suggest other than to have a test or two that calls both `failure()` and `success()` one or more times and asserts that we don't blow up. "Undefined reporting behavior" seems okay but segfaulting doesn't. I don't think this will segfault or anything but a refactor could change that.

Reply:
> I get nervous whenever I see `std::move` on an instance-variable that doesn't result in the whole instance being in a moved-from state. I'm not sure if you need `operator=(&&)` as such (and it would force you to make some fields non-`const`).

I believe an instance-variable being in a moved-from state but the whole instance not being in one is how the `extract()` methods on the `bsoncxx::builder::basic::array` and `bsoncxx::builder ::basic::document` behave. The idea is that you could `std::move()` into them again (using the move assignment operator) and put them back into a valid state again.

> I'm not sure what to suggest other than to have a test or two that calls both `failure()` and `success()` one or more times and asserts that we don't blow up. "Undefined reporting behavior" seems okay but segfaulting doesn't. I don't think this will segfault or anything but a refactor could change that.

I'd rather genny crash loudly in cases of programmer error like this. While Catch2 doesn't support death tests, we could potentially [lift the functionality from the server codebase](https://github.com/mongodb/mongo/blob/r4.1.8/src/mongo/unittest/death_test.cpp).